docs/file-conventions/root.md

```diff
@@ -146,6 +146,24 @@ export function ErrorBoundary() {
 }
 ```
 
+**A note on `useLoaderData`in the `Layout` Component**
+
+`useLoaderData` is not permitted to be used in `ErrorBoundary` components because it is intended for the happy-path route rendering, and it's typings have a built-in assumption that the `loader` ran successfully and returned something. That assumption doesn't hold in an `ErrorBoundary` because it could hve been the `loader` that threw and triggered the boundary! In order to access loader data in `ErrorBoundary`'s, you can use `useRouteLoaderData` which accounts for the loader data being potentially `undefined`.
+
+Because your `Layout` component is used in both success and error flows, this same restriction holds. If you need to fork logic in your `Layout` depending on if it was a successful request or not, you can use `useRouteLoaderData("root")`:
+
+```tsx
+export function Layout({
+  children,
+}: {
+  children: React.ReactNode;
+}) {
+  const data = useRouteLoaderData("root");
+  const error = useRouteError();
+  return <html>...</html>;
+}
```

Comment:
Only suggestion I have is including a little more example of what you would do with this information. Open to a better example, just figure this might help folks understand the purpose a bit better

```suggestion
export function Layout({
  children,
}: {
  children: React.ReactNode;
}) {
  const data = useRouteLoaderData("root");
  const error = useRouteError();
  return (
    <html>
      <head>{/* */}</head>
      <body>
        {data ? (
          <RouteLayout>{children}</RouteLayout>
        ) : (
          <ErrorLayout>{children}</ErrorLayout>
        )}
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}
```

Reply:
Adding analytics using an env variable from the root loader is a good use case

If `useRouteLoaderData("root")` returns a value, you render `<Analytics token={loaderData.env.analytics} />`, otherwise you don't

Reply:
It's tricky because there are other ways to do both of these.  I prefer I prefer doing the dynamic stylesheet via `meta`, and the `<Analytics>` component could probably be done via the root component - but I do understand folks will want to put stuff that feels "global" like this in the `Layout`.  I updated with the Analytics example and the CSS custom property example from the linked discussion. 👍

Reply:
@brophdawg11 thank you very much! I note that you recommend `meta` for dynamic stylesheet `<link>`s but the doc precisely discourages the `meta` export for stylesheets in favor ot the `links` one instead (which can't depend on the data):

> For asset links like stylesheets and favicons, you should use the [links export](https://remix.run/docs/en/main/route/links) instead.

 I always want to follow the official recommendations but this use case seems to be in a grey area. Could you maybe add to the docs that dynamic links are an exception of what is tolerated in the `meta` export? It's not ideal though because it's not the intention of this export (made for SEO). Any suggestion? :)

Reply:
The `link` here depended on loader data so it could not be done in `links` and would have to be done in `meta` (or could also be done in `Layout`)